src/backend/components/py.pi/caliopen_pi/qualifiers/base.py

```diff
@@ -63,11 +65,18 @@ def get_participant(self, message, participant):
         p.type = participant.type
         p.label = participant.label
         p.protocol = message.message_protocol
+        lookup = ParticipantLookup.get_or_create(self.user, p.address,
+                                                 p.protocol)
+        p.participant_id = lookup.participant_id
         log.debug('Will lookup contact {} for user {}'.
                   format(participant.address, self.user.user_id))
         c = Contact.lookup(self.user, participant.address)
         if c:
             p.contact_ids = [c.contact_id]
+            ParticipantLookup.get_or_create(self.user,
+                                            c.contact_id,
+                                            'contact',
+                                            p.participant_id)
```

Comment:
algo is incomplete because it doesn't handle the case where a contact already exists for participant. In this case, algo must NOT create a new participant_id.